File: server/modules/ai/agent.js
const { TOOL_SCHEMAS, TOOL_FUNCTIONS } = require('./tools');

const OLLAMA_URL = 'http://localhost:11434/api/chat';
const MODEL = 'qwen2.5:7b-instruct';
const MAX_ITERATIONS = 6;

const SYSTEM_PROMPT = `You are SiteLine AI, an intelligent assistant for enterprise asset management. You have access to a real database with 500+ assets, 50+ employees, and 12 departments.

When answering questions:
1. Think about which tools to use
2. Call the appropriate tools to gather data
3. Analyze the results
4. If you need more info, call more tools
5. Provide specific, data-driven answers with numbers and names

Available asset categories: Laptops, Desktops, Monitors, Printers, Servers, Network Equipment, Phones, Tablets, Furniture, Vehicles, Lab Equipment, Safety Equipment, AV Equipment, Power Tools, HVAC Systems
Asset statuses: Available, Allocated, In Maintenance, Retired, Lost
Maintenance priorities: Critical, High, Medium, Low
Allocation statuses: Active, Returned, Overdue

Always be helpful, concise, and reference specific data from your tool calls.`;

async function runAgent(userMessage, conversationHistory = []) {
  const messages = [
    { role: 'system', content: SYSTEM_PROMPT },
    ...conversationHistory,
    { role: 'user', content: userMessage }
  ];
  const trace = [];

  for (let i = 0; i < MAX_ITERATIONS; i++) {
    let response;
    try {
      const res = await fetch(OLLAMA_URL, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ model: MODEL, messages, tools: TOOL_SCHEMAS, stream: false }),
        signal: AbortSignal.timeout(120000)
      });
      response = await res.json();
    } catch (err) {
      throw new Error(`Ollama connection failed: ${err.message}`);
    }

    const msg = response.message;
    if (!msg) throw new Error('No response from Ollama');
    messages.push(msg);

    if (!msg.tool_calls || msg.tool_calls.length === 0) {
      return { answer: msg.content || 'No response generated.', trace };
    }

    for (const toolCall of msg.tool_calls) {
      const fnName = toolCall.function?.name;
      const fn = TOOL_FUNCTIONS[fnName];
      let result;
      try {
        result = fn ? await fn(toolCall.function.arguments || {}) : { error: `Unknown tool: ${fnName}` };
      } catch (e) {
        result = { error: e.message };
      }
      trace.push({ tool: fnName, arguments: toolCall.function?.arguments || {}, result });
      messages.push({ role: 'tool', content: JSON.stringify(result) });
    }
  }

  // Force final answer after max iterations
  messages.push({ role: 'user', content: 'Please provide your final answer based on the data gathered so far.' });
  try {
    const res = await fetch(OLLAMA_URL, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ model: MODEL, messages, stream: false }),
      signal: AbortSignal.timeout(120000)
    });
    const final = await res.json();
    return { answer: final.message?.content || 'Unable to generate response.', trace };
  } catch {
    return { answer: 'Max iterations reached. Here is what I found so far.', trace };
  }
}

module.exports = { runAgent };

File: agent-test/backend/agent.py
"""
AssetFlow Agent - Agentic loop with Ollama tool-calling.
"""
import requests
import json
from backend.tools import TOOL_SCHEMAS, TOOL_FUNCTIONS

OLLAMA_URL = 'http://localhost:11434/api/chat'
MODEL = 'qwen2.5:7b-instruct'
MAX_ITERATIONS = 6

SYSTEM_PROMPT = """You are AssetFlow Agent, an intelligent assistant for manufacturing asset management. You have access to a real database of manufacturing assets, employees, maintenance records, and bookings across multiple plants.

When a user asks a question:
1. Think about which tool(s) you need to answer it
2. Call the appropriate tool(s)
3. Analyze the results
4. If you need more information, call additional tools
5. Provide a clear, helpful answer based on the data

Always be specific with numbers and names from the actual data. If a question requires multiple lookups, do them sequentially.

Available asset categories: CNC Machine, Forklift, Conveyor Belt, Compressor, Generator, Laptop, Meeting Room, Vehicle, Welding Station, 3D Printer.
Asset statuses: Available, Allocated, Under Maintenance, Reserved, Retired.
Plants: Detroit Manufacturing Hub (Detroit), Austin Tech Center (Austin), Chicago Distribution Center (Chicago)."""


def run_agent(user_message: str, conversation_history: list, db_path: str) -> dict:
    """
    Execute the agentic loop:
    1. Send user message + tools to the LLM
    2. If the LLM requests tool calls, execute them and feed results back
    3. Repeat until the LLM gives a final text answer or MAX_ITERATIONS is hit
    """
    trace = []
    messages = [
        {'role': 'system', 'content': SYSTEM_PROMPT}
    ]

    # Add conversation history
    for msg in conversation_history:
        messages.append(msg)

    messages.append({'role': 'user', 'content': user_message})

    for iteration in range(MAX_ITERATIONS):
        # Call Ollama
        payload = {
            'model': MODEL,
            'messages': messages,
            'tools': TOOL_SCHEMAS,
            'stream': False,
        }

        try:
            response = requests.post(OLLAMA_URL, json=payload, timeout=120)
            response.raise_for_status()
        except requests.RequestException as e:
            return {
                'answer': f'Error communicating with Ollama: {str(e)}',
                'trace': trace,
            }

        result = response.json()
        assistant_message = result['message']
        messages.append(assistant_message)

        # Check if model wants to call tools
        if 'tool_calls' not in assistant_message or not assistant_message['tool_calls']:
            # Model gave a final text answer
            return {
                'answer': assistant_message.get('content', ''),
                'trace': trace,
            }

        # Execute each tool call
        for tool_call in assistant_message['tool_calls']:
            func_name = tool_call['function']['name']
            func_args = tool_call['function'].get('arguments', {})

            print(f"[Agent] Iteration {iteration + 1} | Tool call: {func_name}({json.dumps(func_args)})")

            # Execute the function
            if func_name in TOOL_FUNCTIONS:
                try:
                    tool_result = TOOL_FUNCTIONS[func_name](db_path, **func_args)
                except Exception as e:
                    tool_result = {'error': str(e)}
            else:
                tool_result = {'error': f'Unknown tool: {func_name}'}

            result_str = json.dumps(tool_result, default=str)
            print(f"[Agent] Result preview: {result_str[:500]}")

            trace.append({
                'tool': func_name,
                'arguments': func_args,
                'result': tool_result,
            })

            # Add tool result to messages
            messages.append({
                'role': 'tool',
                'content': result_str,
            })

    # Max iterations reached - force a final answer
    messages.append({
        'role': 'user',
        'content': 'Please provide your final answer based on the information gathered so far.',
    })
    payload = {
        'model': MODEL,
        'messages': messages,
        'stream': False,
    }
    try:
        response = requests.post(OLLAMA_URL, json=payload, timeout=120)
        response.raise_for_status()
        result = response.json()
        return {
            'answer': result['message'].get(
                'content',
                'Max iterations reached. Here is what I found so far based on the tool calls above.'
            ),
            'trace': trace,
        }
    except requests.RequestException as e:
        return {
            'answer': f'Max iterations reached. Error getting final answer: {str(e)}',
            'trace': trace,
        }
